In [ ]:
import pandas as pd
import requests 
import os
from dotenv import load_dotenv, find_dotenv
import base64
import json
import numpy as np
import shutil

In [73]:
with open('albums.json') as f:
    albums_data = json.load(f)

with open('all_tracks.json') as g:
    all_tracks_data = json.load(g)

with open('top_tracks.json') as h:
    top_tracks_data = json.load(h)


In [74]:
df_all_tracks = pd.DataFrame(all_tracks_data)

df_albums = pd.DataFrame(albums_data)

df_top_tracks = pd.DataFrame(top_tracks_data)


In [75]:
df_all_tracks.shape


(530, 16)

In [76]:
df_all_tracks.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 530 entries, 0 to 529
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   artists            530 non-null    object
 1   available_markets  530 non-null    object
 2   disc_number        530 non-null    int64 
 3   duration_ms        530 non-null    int64 
 4   explicit           530 non-null    bool  
 5   external_urls      530 non-null    object
 6   href               530 non-null    object
 7   id                 530 non-null    object
 8   name               530 non-null    object
 9   preview_url        0 non-null      object
 10  track_number       530 non-null    int64 
 11  type               530 non-null    object
 12  uri                530 non-null    object
 13  is_local           530 non-null    bool  
 14  album_name         530 non-null    object
 15  restrictions       8 non-null      object
dtypes: bool(2), int64(3), object(11)
memory usag

In [77]:
df_all_tracks.head()

artists  \
0  [{'external_urls': {'spotify': 'https://open.s...   
1  [{'external_urls': {'spotify': 'https://open.s...   
2  [{'external_urls': {'spotify': 'https://open.s...   
3  [{'external_urls': {'spotify': 'https://open.s...   
4  [{'external_urls': {'spotify': 'https://open.s...   

                                   available_markets  disc_number  \
0  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
1  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
2  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
3  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
4  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   

   duration_ms  explicit                                      external_urls  \
0       649500     False  {'spotify': 'https://open.spotify.com/track/7E...   
1       599000     False  {'spotify': 'https://open.spotify.com/track/5v...   
2       282000     False  {'spotify': 'https://open.spotify.com/track/2t...   
3       228000     False  {'spotify': 'https://open.spotify.com/track/2x...   
4       224500     False  {'spotify': 'https://open.spotify.com/track/5X...   

                                                href                      id  \
0  https://api.spotify.com/v1/tracks/7EHLirYEZgOs...  7EHLirYEZgOsN5aLoBvcsh   
1  https://api.spotify.com/v1/tracks/5vBIR7QYQolI...  5vBIR7QYQolI2YxTM2GW19   
2  https://api.spotify.com/v1/tracks/2ttmDbMAbrxy...  2ttmDbMAbrxyTQJlzyCLxq   
3  https://api.spotify.com/v1/tracks/2xwODdnX2fbB...  2xwODdnX2fbBu2XxInQqEa   
4  https://api.spotify.com/v1/tracks/5XTDwwTFy07r...  5XTDwwTFy07rUDcGS9fvUY   

                                             name preview_url  track_number  \
0       no stillson 6 cirk [Houston, TX 12.17.16]        None             1   
1  no stillson 6 cirk mix2 [Houston, TX 12.17.16]        None             2   
2                42DIMENSIT3 e3 [London 03.06.17]        None             3   
3       MT1T1 bedroom microtune [London 03.06.17]        None             4   
4                    T18A pole1 [London 03.06.17]        None             5   

    type                                   uri  is_local  \
0  track  spotify:track:7EHLirYEZgOsN5aLoBvcsh     False   
1  track  spotify:track:5vBIR7QYQolI2YxTM2GW19     False   
2  track  spotify:track:2ttmDbMAbrxyTQJlzyCLxq     False   
3  track  spotify:track:2xwODdnX2fbBu2XxInQqEa     False   
4  track  spotify:track:5XTDwwTFy07rUDcGS9fvUY     False   

                                album_name restrictions  
0  Music From The Merch Desk (2016 - 2023)          NaN  
1  Music From The Merch Desk (2016 - 2023)          NaN  
2  Music From The Merch Desk (2016 - 2023)          NaN  
3  Music From The Merch Desk (2016 - 2023)          NaN  
4  Music From The Merch Desk (2016 - 2023)          NaN

In [78]:
df_all_tracks.isnull()

artists  available_markets  disc_number  duration_ms  explicit  \
0      False              False        False        False     False   
1      False              False        False        False     False   
2      False              False        False        False     False   
3      False              False        False        False     False   
4      False              False        False        False     False   
..       ...                ...          ...          ...       ...   
525    False              False        False        False     False   
526    False              False        False        False     False   
527    False              False        False        False     False   
528    False              False        False        False     False   
529    False              False        False        False     False   

     external_urls   href     id   name  preview_url  track_number   type  \
0            False  False  False  False         True         False  False   
1            False  False  False  False         True         False  False   
2            False  False  False  False         True         False  False   
3            False  False  False  False         True         False  False   
4            False  False  False  False         True         False  False   
..             ...    ...    ...    ...          ...           ...    ...   
525          False  False  False  False         True         False  False   
526          False  False  False  False         True         False  False   
527          False  False  False  False         True         False  False   
528          False  False  False  False         True         False  False   
529          False  False  False  False         True         False  False   

       uri  is_local  album_name  restrictions  
0    False     False       False          True  
1    False     False       False          True  
2    False     False       False          True  
3    False     False       False          True  
4    False     False       False          True  
..     ...       ...         ...           ...  
525  False     False       False          True  
526  False     False       False          True  
527  False     False       False          True  
528  False     False       False          True  
529  False     False       False          True  

[530 rows x 16 columns]

# COMPREENSÃO DOS DADOS

* artist: nested dict contendo diversos links sobre o artista, como a URI (identificador universal) do artista além da URL que leva o usuário diretamente para a página do Spotify. O nome do artista e seu id, todos em forma de key values. Poderá ser descartado pois todas essas informações são redundantes no dataframe (já temos coluna contendo a URL, seu id e seu nome)

* available_markets: Array de países que permitem a reprodução dessa música específica

* disc_number: Número de discos que o álbum que a música está contém (Ex: Souvlaki de Slowdive possuem dois discos, mesmo sendo 1 álbum)

* duration_ms: Duração da música em milissegundos

* explicit: Booleano que indica se a música possui conteúdo explícito (contém conteúdo que o espectador precisa ler com percepção de grau de violência, como violas, danças, ou música de violência)

* external_urls: Como dito anteriormente, contém a url que leva à pagina web do artista

* href: Contém o link que leva até o endpoint da API, pode ser descartado

* id: Id único da música

* name: nome da música

* preview_url: contém uma prévia em formato mp3 de 30 segundos da música, o tipo de dado é object e o python traduz objetos vazios como 'None' e não 'Nan'. Também pode ser deletado

* track_number: Número da música dentro do álbum

* type: O tipo da música, sempre 'track'

* uri: URI que identifica a música unicamente

* is_local: Se a faixa é ou não de um arquivo local.

* album_name: Nome do álbum

* restrictions: Se há alguma restrição que impeça o usuário de acessar a música, pela minha avaliação, todas as músicas NÃO estão restritas, essa coluna também será deletada


In [79]:
df_all_tracks_clean = df_all_tracks.drop(['artists','disc_number','restrictions','preview_url','href','type','is_local'], axis=1)

df_all_tracks_clean.iloc[310:320]



available_markets  duration_ms  explicit  \
310  [AR, AU, AT, BE, BO, BR, BG, CL, CO, CY, CZ, D...       159146     False   
311  [AR, AU, AT, BE, BO, BR, BG, CL, CO, CY, CZ, D...       190066     False   
312  [AR, AU, AT, BE, BO, BR, BG, CL, CO, CY, CZ, D...       257226     False   
313  [AR, AU, AT, BE, BO, BR, BG, CL, CO, CY, CZ, D...       315506     False   
314  [AR, AU, AT, BE, BO, BR, BG, CL, CO, CY, CZ, D...       295680     False   
315  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       114626     False   
316  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       206160     False   
317  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       150648     False   
318  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       187520     False   
319  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       268000     False   

                                         external_urls  \
310  {'spotify': 'https://open.spotify.com/track/2P...   
311  {'spotify': 'https://open.spotify.com/track/2s...   
312  {'spotify': 'https://open.spotify.com/track/4e...   
313  {'spotify': 'https://open.spotify.com/track/1E...   
314  {'spotify': 'https://open.spotify.com/track/0m...   
315  {'spotify': 'https://open.spotify.com/track/0r...   
316  {'spotify': 'https://open.spotify.com/track/2W...   
317  {'spotify': 'https://open.spotify.com/track/6m...   
318  {'spotify': 'https://open.spotify.com/track/1l...   
319  {'spotify': 'https://open.spotify.com/track/62...   

                         id  \
310  2PBfwF7Xghmmoe2gfW6iiH   
311  2s6NUgm38Nx4PQcHHAe4T9   
312  4e9t6iY0JWqRaFaImaM9LD   
313  1EylkASjTjZ6qRUWBpFHJl   
314  0ms3DnVh1X838J26L6SquU   
315  0rjIwXjUmd6OBIZKMq3gwQ   
316  2WivpjH03Usq17IwgEZIwb   
317  6m2fEpTO7uLzSfaw6M6rww   
318  1lzcCXn7i2P269pGbguxd6   
319  62gGUfA8mlHSzTChoP3Mnu   

                                                  name  track_number  \
310                                       Come on Feet            16   
311                                  Sugar Coated Lies            17   
312                                   Witness (1 Hope)            18   
313                               Sponsored by Destiny            19   
314                                        Temple Head            20   
315  Vivaldi Variation (Arr. for Piano from Concert...             1   
316                         Memories of a Promenade II             2   
317  Schumann Variation (on a Theme from Piano Conc...             3   
318               Gymnopédie No. 1: Lent et douloureux             4   
319  Rinaldo, HWV 7: Lascia ch'io pianga (Arr. for ...             5   

                                      uri  \
310  spotify:track:2PBfwF7Xghmmoe2gfW6iiH   
311  spotify:track:2s6NUgm38Nx4PQcHHAe4T9   
312  spotify:track:4e9t6iY0JWqRaFaImaM9LD   
313  spotify:track:1EylkASjTjZ6qRUWBpFHJl   
314  spotify:track:0ms3DnVh1X838J26L6SquU   
315  spotify:track:0rjIwXjUmd6OBIZKMq3gwQ   
316  spotify:track:2WivpjH03Usq17IwgEZIwb   
317  spotify:track:6m2fEpTO7uLzSfaw6M6rww   
318  spotify:track:1lzcCXn7i2P269pGbguxd6   
319  spotify:track:62gGUfA8mlHSzTChoP3Mnu   

                                         album_name  
310  A to Z: Bestival 2008: Compiled by Rob Da Bank  
311  A to Z: Bestival 2008: Compiled by Rob Da Bank  
312  A to Z: Bestival 2008: Compiled by Rob Da Bank  
313  A to Z: Bestival 2008: Compiled by Rob Da Bank  
314  A to Z: Bestival 2008: Compiled by Rob Da Bank  
315       Piano Classics - Relaxing Classical Music  
316       Piano Classics - Relaxing Classical Music  
317       Piano Classics - Relaxing Classical Music  
318       Piano Classics - Relaxing Classical Music  
319       Piano Classics - Relaxing Classical Music

In [80]:
x = df_all_tracks_clean['available_markets'] #verifica quantos países permitem a reprodução da faixa na primeira linha

print(len(x)) 


# Conjunto de siglas totais (códigos de países)
siglas_totais = {
    'AF', 'AL', 'DZ', 'AS', 'AD', 'AO', 'AI', 'AQ', 'AG', 'AR', 'AM', 'AW', 'AU', 'AT', 'AZ', 
    'BS', 'BH', 'BD', 'BB', 'BY', 'BE', 'BZ', 'BJ', 'BM', 'BT', 'BO', 'BA', 'BW', 'BR', 'IO', 
    'BN', 'BG', 'BF', 'BI', 'KH', 'CM', 'CA', 'CV', 'KY', 'CF', 'TD', 'CL', 'CN', 'CX', 'CC', 
    'CO', 'KM', 'CD', 'CG', 'CK', 'CR', 'CI', 'HR', 'CU', 'CY', 'CZ', 'DK', 'DJ', 'DM', 'DO', 
    'EC', 'EG', 'SV', 'GQ', 'ER', 'EE', 'SZ', 'ET', 'FK', 'FO', 'FJ', 'FI', 'FR', 'GF', 'PF', 
    'TF', 'GA', 'GM', 'GE', 'DE', 'GH', 'GI', 'GR', 'GL', 'GD', 'GP', 'GU', 'GT', 'GG', 'GN', 
    'GW', 'GY', 'HT', 'HM', 'VA', 'HN', 'HK', 'HU', 'IS', 'IN', 'ID', 'IR', 'IQ', 'IE', 'IM', 
    'IL', 'IT', 'JM', 'JP', 'JE', 'JO', 'KZ', 'KE', 'KI', 'KP', 'KR', 'KW', 'KG', 'LA', 'LV', 
    'LB', 'LS', 'LR', 'LY', 'LI', 'LT', 'LU', 'MO', 'MG', 'MW', 'MY', 'MV', 'ML', 'MT', 'MH', 
    'MQ', 'MR', 'MU', 'YT', 'MX', 'FM', 'MD', 'MC', 'MN', 'ME', 'MS', 'MA', 'MZ', 'MM', 'NA', 
    'NR', 'NP', 'NL', 'NC', 'NZ', 'NI', 'NE', 'NG', 'NU', 'NF', 'MK', 'MP', 'NO', 'OM', 'PK', 
    'PW', 'PS', 'PA', 'PG', 'PY', 'PE', 'PH', 'PN', 'PL', 'PT', 'PR', 'QA', 'RE', 'RO', 'RU', 
    'RW', 'BL', 'SH', 'KN', 'LC', 'MF', 'PM', 'VC', 'WS', 'SM', 'ST', 'SA', 'SN', 'RS', 'SC', 
    'SL', 'SG', 'SX', 'SK', 'SI', 'SB', 'SO', 'ZA', 'GS', 'SS', 'ES', 'LK', 'SD', 'SR', 'SJ', 
    'SE', 'CH', 'SY', 'TW', 'TJ', 'TZ', 'TH', 'TL', 'TG', 'TK', 'TO', 'TT', 'TN', 'TR', 'TM', 
    'TC', 'TV', 'UG', 'UA', 'AE', 'GB', 'US', 'UY', 'UZ', 'VU', 'VE', 'VN', 'WF', 'EH', 'YE', 
    'ZM', 'ZW'
}

def VerificarSiglasFaltantes(siglas_totais, x):
    for i, linhas in enumerate(x.to_numpy()):
        linha = set(linhas)  
        siglas_faltantes = siglas_totais - linha  
        print(f"Linha {i+1}:{len(siglas_faltantes)} mercados não aceitam a comercialização das musicas do Aphex Twin, sendo eles: {siglas_faltantes}.")




#https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2 -> Lista de códigos de países permitidos. Fonte: https://developer.spotify.com/documentation/web-api/reference/get-an-album
# total de países listados: 242

530


#Pela análise, a maioria das músicas do Aphex Twin podem ser comercializadas em grande parte do mundo, com exceção das linhas 296 até 315. 

- Investigando mais a fundo, percebi duas coisas: 1. A nossa "gambiarra" de extrair todas as músicas do Aphex levava em consideração todos os álbums que o Aphex participou pelo menos 1 vez. Ou seja, essas linhas (tirando a 295) não tinham um unico dedo do Aphex, nem nos créditos.
2. Todos os códigos seguem um padrão, ficando enre 59 e 62, como não era uma diferença muito grande, optei por não explorar mais a fundo.
- Então, a coluna "artist", que continha o nome do artista, não é completamente redundante, devemos agora extrair o dicionario para verificar quais musicas de fato são do Aphex
3. Por limitações da API, a coluna "disc_number" não está contailizando corretamente os discos, logo será removida.


In [81]:
name = df_all_tracks['artists']
df_name= pd.DataFrame(name)

dict_artist = df_name.to_dict()

nome = dict_artist['artists'][0][0]['name']
uri = dict_artist['artists'][0][0]['uri']
external_urls =  dict_artist['artists'][0][0]['external_urls']['spotify']
dict_artist['artists'][0][0]['external_urls']['spotify']
df_all_tracks_clean['external_urls'] = dict_artist['artists'][0][0]['external_urls']['spotify']




Mesmo extraindo os dados da coluna artist, a API reconhece como sendo do Aphex, acredito que o melhor jeito de tratar esses dados será no SQL através de um JOIN entre a tabela ALBUMS e ALL_TRACKS.

In [82]:
df_all_tracks_clean.drop(['uri'], axis=1)

df_all_tracks_processed = df_all_tracks_clean.drop(['uri'], axis=1)

df_all_tracks_processed.head()

available_markets  duration_ms  explicit  \
0  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       649500     False   
1  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       599000     False   
2  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       282000     False   
3  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       228000     False   
4  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       224500     False   

                                       external_urls                      id  \
0  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  7EHLirYEZgOsN5aLoBvcsh   
1  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  5vBIR7QYQolI2YxTM2GW19   
2  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  2ttmDbMAbrxyTQJlzyCLxq   
3  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  2xwODdnX2fbBu2XxInQqEa   
4  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  5XTDwwTFy07rUDcGS9fvUY   

                                             name  track_number  \
0       no stillson 6 cirk [Houston, TX 12.17.16]             1   
1  no stillson 6 cirk mix2 [Houston, TX 12.17.16]             2   
2                42DIMENSIT3 e3 [London 03.06.17]             3   
3       MT1T1 bedroom microtune [London 03.06.17]             4   
4                    T18A pole1 [London 03.06.17]             5   

                                album_name  
0  Music From The Merch Desk (2016 - 2023)  
1  Music From The Merch Desk (2016 - 2023)  
2  Music From The Merch Desk (2016 - 2023)  
3  Music From The Merch Desk (2016 - 2023)  
4  Music From The Merch Desk (2016 - 2023)

Para o tratamento de df_albums. Terá de ser feito algo mais rigoroso, levando em consideração principalmente qu esão menos linhas, permitindo uma análise mais aprofundada, Vou deixar no Dataframe apenas os albums, singles, EPs que sejam de fato do Aphex Twin, e não apenas participação. Além disso, vou aplicar algumas das soluções ja feitas antes.

In [83]:
df_albums.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   album_type              50 non-null     object
 1   total_tracks            50 non-null     int64 
 2   available_markets       50 non-null     object
 3   external_urls           50 non-null     object
 4   href                    50 non-null     object
 5   id                      50 non-null     object
 6   images                  50 non-null     object
 7   name                    50 non-null     object
 8   release_date            50 non-null     object
 9   release_date_precision  50 non-null     object
 10  type                    50 non-null     object
 11  uri                     50 non-null     object
 12  artists                 50 non-null     object
 13  album_group             50 non-null     object
dtypes: int64(1), object(13)
memory usage: 5.6+ KB


In [84]:
df_albums.head()

album_type  total_tracks                                  available_markets  \
0      album            38  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
1      album            27  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
2      album            10  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
3      album            12  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
4      album            30  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   

                                       external_urls  \
0  {'spotify': 'https://open.spotify.com/album/72...   
1  {'spotify': 'https://open.spotify.com/album/7g...   
2  {'spotify': 'https://open.spotify.com/album/58...   
3  {'spotify': 'https://open.spotify.com/album/4m...   
4  {'spotify': 'https://open.spotify.com/album/2o...   

                                                href                      id  \
0  https://api.spotify.com/v1/albums/72pxRHG1VNIc...  72pxRHG1VNIcIibziKpH03   
1  https://api.spotify.com/v1/albums/7gpblNxStuPC...  7gpblNxStuPCUouPvlyf2i   
2  https://api.spotify.com/v1/albums/58H1ArDcd3bU...  58H1ArDcd3bUzOK0cn31lD   
3  https://api.spotify.com/v1/albums/4mwrMLVKo940...  4mwrMLVKo940qLFXEIef4w   
4  https://api.spotify.com/v1/albums/2oJo7cB45gMV...  2oJo7cB45gMVuRsaWNwDq2   

                                              images  \
0  [{'url': 'https://i.scdn.co/image/ab67616d0000...   
1  [{'url': 'https://i.scdn.co/image/ab67616d0000...   
2  [{'url': 'https://i.scdn.co/image/ab67616d0000...   
3  [{'url': 'https://i.scdn.co/image/ab67616d0000...   
4  [{'url': 'https://i.scdn.co/image/ab67616d0000...   

                                                name release_date  \
0            Music From The Merch Desk (2016 - 2023)   2024-12-17   
1  Selected Ambient Works Volume II (Expanded Edi...   2024-10-04   
2                       Digeridoo (Expanded Edition)   2024-05-31   
3                                               Syro   2014-09-22   
4                                             Drukqs   2001-10-22   

  release_date_precision   type                                   uri  \
0                    day  album  spotify:album:72pxRHG1VNIcIibziKpH03   
1                    day  album  spotify:album:7gpblNxStuPCUouPvlyf2i   
2                    day  album  spotify:album:58H1ArDcd3bUzOK0cn31lD   
3                    day  album  spotify:album:4mwrMLVKo940qLFXEIef4w   
4                    day  album  spotify:album:2oJo7cB45gMVuRsaWNwDq2   

                                             artists album_group  
0  [{'external_urls': {'spotify': 'https://open.s...       album  
1  [{'external_urls': {'spotify': 'https://open.s...       album  
2  [{'external_urls': {'spotify': 'https://open.s...       album  
3  [{'external_urls': {'spotify': 'https://open.s...       album  
4  [{'external_urls': {'spotify': 'https://open.s...       album

In [85]:
df_albums.shape

(50, 14)

# COMPREENSÃO DOS DADOS

* album_type = variável categórica podendo ser Album, Single (single pode ser um grupo de track), Compilados e EPs, estarei deletando essa coluna pois para nossa análise, tudo será tratado como álbum, além da redundância ja que o nome do projeto está em outras colunas.
* total_tracks = número de tracks para cada album
* available_market = Array de países que permitem a reprodução dessa música específica
* external_url = Como dito anteriormente, contém a url que leva à pagina web do artista
* href = Contém o link que leva até o endpoint da API, pode ser descartado
* id = id único do álbum
* images = arquivo contendo a capa do álbum 
* name = Título do Álbum
* release_date = Data de lançamento do Álbum
* release_date_precision = A precisão com que release_date valor é conhecido.
* type = o tipo (sempre será álbum). Será descartado por ser uma informação redundante
* uri = a uri do álbum
* artists = nested dict contendo diversos links sobre o artista, como a URI (identificador universal) do artista além da URL que leva o usuário diretamente para a página do Spotify. O nome do artista e seu id, todos em forma de key values. Poderá ser descartado pois todas essas informações são redundantes no dataframe (já temos coluna contendo a URL, seu id e seu nome)
* album_group = Este campo descreve a relação entre o artista e o álbum. ( Álbum, Feat, Single)

In [86]:
df_albums['album_group'] == 'appears_on'

df_albums_Aphex = df_albums[df_albums['album_group'] != 'appears_on']

df_albums_Aphex.shape



(31, 14)

In [87]:
df_albums_Aphex_clean = df_albums_Aphex.drop(['href','type','release_date_precision','images'], axis=1)

df_albums_Aphex_clean.head()

album_type  total_tracks                                  available_markets  \
0      album            38  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
1      album            27  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
2      album            10  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
3      album            12  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
4      album            30  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   

                                       external_urls                      id  \
0  {'spotify': 'https://open.spotify.com/album/72...  72pxRHG1VNIcIibziKpH03   
1  {'spotify': 'https://open.spotify.com/album/7g...  7gpblNxStuPCUouPvlyf2i   
2  {'spotify': 'https://open.spotify.com/album/58...  58H1ArDcd3bUzOK0cn31lD   
3  {'spotify': 'https://open.spotify.com/album/4m...  4mwrMLVKo940qLFXEIef4w   
4  {'spotify': 'https://open.spotify.com/album/2o...  2oJo7cB45gMVuRsaWNwDq2   

                                                name release_date  \
0            Music From The Merch Desk (2016 - 2023)   2024-12-17   
1  Selected Ambient Works Volume II (Expanded Edi...   2024-10-04   
2                       Digeridoo (Expanded Edition)   2024-05-31   
3                                               Syro   2014-09-22   
4                                             Drukqs   2001-10-22   

                                    uri  \
0  spotify:album:72pxRHG1VNIcIibziKpH03   
1  spotify:album:7gpblNxStuPCUouPvlyf2i   
2  spotify:album:58H1ArDcd3bUzOK0cn31lD   
3  spotify:album:4mwrMLVKo940qLFXEIef4w   
4  spotify:album:2oJo7cB45gMVuRsaWNwDq2   

                                             artists album_group  
0  [{'external_urls': {'spotify': 'https://open.s...       album  
1  [{'external_urls': {'spotify': 'https://open.s...       album  
2  [{'external_urls': {'spotify': 'https://open.s...       album  
3  [{'external_urls': {'spotify': 'https://open.s...       album  
4  [{'external_urls': {'spotify': 'https://open.s...       album

In [88]:
dict_album = df_albums_Aphex_clean.to_dict()
nome = dict_album['artists'][0][0]['name']
uri = dict_album['artists'][0][0]['uri']
external_urls =  dict_album['artists'][0][0]['external_urls']['spotify']

df_albums_Aphex_clean['external_urls'] = dict_album['artists'][0][0]['external_urls']['spotify']
df_albums_Aphex_clean['uri'] = uri = dict_album['artists'][0][0]['uri']

df_albums_Aphex_clean.drop(['artists'], axis=1)

df_albums_Aphex_processed = df_albums_Aphex_clean.drop(['artists'], axis=1)

df_albums_Aphex_processed.head()

album_type  total_tracks                                  available_markets  \
0      album            38  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
1      album            27  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
2      album            10  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
3      album            12  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   
4      album            30  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...   

                                       external_urls                      id  \
0  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  72pxRHG1VNIcIibziKpH03   
1  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  7gpblNxStuPCUouPvlyf2i   
2  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  58H1ArDcd3bUzOK0cn31lD   
3  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  4mwrMLVKo940qLFXEIef4w   
4  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  2oJo7cB45gMVuRsaWNwDq2   

                                                name release_date  \
0            Music From The Merch Desk (2016 - 2023)   2024-12-17   
1  Selected Ambient Works Volume II (Expanded Edi...   2024-10-04   
2                       Digeridoo (Expanded Edition)   2024-05-31   
3                                               Syro   2014-09-22   
4                                             Drukqs   2001-10-22   

                                     uri album_group  
0  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG       album  
1  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG       album  
2  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG       album  
3  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG       album  
4  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG       album

In [89]:
df_top_tracks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   album              10 non-null     object
 1   artists            10 non-null     object
 2   available_markets  10 non-null     object
 3   disc_number        10 non-null     int64 
 4   duration_ms        10 non-null     int64 
 5   explicit           10 non-null     bool  
 6   external_ids       10 non-null     object
 7   external_urls      10 non-null     object
 8   href               10 non-null     object
 9   id                 10 non-null     object
 10  is_local           10 non-null     bool  
 11  is_playable        10 non-null     bool  
 12  name               10 non-null     object
 13  popularity         10 non-null     int64 
 14  preview_url        0 non-null      object
 15  track_number       10 non-null     int64 
 16  type               10 non-null     object
 17  

In [90]:

df_top_tracks.shape

(10, 18)

In [91]:
df_top_tracks.head()

album  \
0  {'album_type': 'album', 'artists': [{'external...   
1  {'album_type': 'album', 'artists': [{'external...   
2  {'album_type': 'album', 'artists': [{'external...   
3  {'album_type': 'album', 'artists': [{'external...   
4  {'album_type': 'album', 'artists': [{'external...   

                                             artists  \
0  [{'external_urls': {'spotify': 'https://open.s...   
1  [{'external_urls': {'spotify': 'https://open.s...   
2  [{'external_urls': {'spotify': 'https://open.s...   
3  [{'external_urls': {'spotify': 'https://open.s...   
4  [{'external_urls': {'spotify': 'https://open.s...   

                                   available_markets  disc_number  \
0  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
1  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
2  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
3  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   
4  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...            1   

   duration_ms  explicit              external_ids  \
0        87466     False  {'isrc': 'GBBPW0100153'}   
1       293751     False  {'isrc': 'BEZ350800022'}   
2       125520     False  {'isrc': 'GBBPW0100148'}   
3       464293     False  {'isrc': 'GBBPW9400034'}   
4       177093     False  {'isrc': 'GBBPW9700077'}   

                                       external_urls  \
0  {'spotify': 'https://open.spotify.com/track/4L...   
1  {'spotify': 'https://open.spotify.com/track/7o...   
2  {'spotify': 'https://open.spotify.com/track/1u...   
3  {'spotify': 'https://open.spotify.com/track/7g...   
4  {'spotify': 'https://open.spotify.com/track/3J...   

                                                href                      id  \
0  https://api.spotify.com/v1/tracks/4LIM4qmpHABu...  4LIM4qmpHABufePRrLWbiM   
1  https://api.spotify.com/v1/tracks/7o2AeQZzfCER...  7o2AeQZzfCERsRmOM86EcB   
2  https://api.spotify.com/v1/tracks/1uaGSDFsLdRe...  1uaGSDFsLdReQgg8p7Obwh   
3  https://api.spotify.com/v1/tracks/7glKwbR1DyuI...  7glKwbR1DyuIuE6XvZvJbQ   
4  https://api.spotify.com/v1/tracks/3JJ4BoL9WVHk...  3JJ4BoL9WVHk4Yye2EGJC7   

   is_local  is_playable        name  popularity preview_url  track_number  \
0     False         True       QKThr          63        None            19   
1     False         True        Xtal          61        None             1   
2     False         True  Avril 14th          59        None             9   
3     False         True          #3          58        None             3   
4     False         True        Flim          56        None             2   

    type                                   uri  
0  track  spotify:track:4LIM4qmpHABufePRrLWbiM  
1  track  spotify:track:7o2AeQZzfCERsRmOM86EcB  
2  track  spotify:track:1uaGSDFsLdReQgg8p7Obwh  
3  track  spotify:track:7glKwbR1DyuIuE6XvZvJbQ  
4  track  spotify:track:3JJ4BoL9WVHk4Yye2EGJC7

# COMPREENSÃO DOS DADOS

* album: O álbum em que a faixa aparece. O objeto album inclui um link em href para informações completas sobre o álbum.
* artists: nested dict contendo diversos links sobre o artista, como a URI (identificador universal) do artista além da URL que leva o usuário diretamente para a página do Spotify. O nome do artista e seu id, todos em forma de key values. Poderá ser descartado pois todas essas informações são redundantes no dataframe (já temos coluna contendo a URL, seu id e seu nome)
* available_markets: Array de países que permitem a reprodução dessa música específica ( extremamente importante pois podemos ver a correlação da distribuição do produto com a popularidade)
* disc_number: Número de discos que o álbum que a música está contém (Ex: Souvlaki de Slowdive possuem dois discos, mesmo sendo 1 álbum)
* duration_ms: Duração da música em milissegundos
* explicit: Booleano que indica se a música possui conteúdo explícito (contém conteúdo que o espectador precisa ler com percepção de grau de violência, como violas, danças, ou música de violência)
* external_ids: Identificadores externos da música (códigos comerciais)
* external_urls: Leva à pagina web da musica no spotify
* href:  Contém o link que leva até o endpoint da API, pode ser descartado
* id:identificador unico da música
* is_local: Se a faixa é ou não de um arquivo local.
* name: Nome da música

* popularity: Um número que representa a popularidade da música. Quanto maior, mais popular a música é.
* preview_url: URL de 30 segundos da musica
* track_number: número da track
*type: Tipo (Ex: Single, Album)
*uri: a uri da música

In [92]:
df_top_tracks_clean = df_top_tracks.drop(['album','href','is_local','is_playable','preview_url','disc_number','uri','external_ids'],axis=1)

df_top_tracks_clean.head()

artists  \
0  [{'external_urls': {'spotify': 'https://open.s...   
1  [{'external_urls': {'spotify': 'https://open.s...   
2  [{'external_urls': {'spotify': 'https://open.s...   
3  [{'external_urls': {'spotify': 'https://open.s...   
4  [{'external_urls': {'spotify': 'https://open.s...   

                                   available_markets  duration_ms  explicit  \
0  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...        87466     False   
1  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       293751     False   
2  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       125520     False   
3  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       464293     False   
4  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       177093     False   

                                       external_urls                      id  \
0  {'spotify': 'https://open.spotify.com/track/4L...  4LIM4qmpHABufePRrLWbiM   
1  {'spotify': 'https://open.spotify.com/track/7o...  7o2AeQZzfCERsRmOM86EcB   
2  {'spotify': 'https://open.spotify.com/track/1u...  1uaGSDFsLdReQgg8p7Obwh   
3  {'spotify': 'https://open.spotify.com/track/7g...  7glKwbR1DyuIuE6XvZvJbQ   
4  {'spotify': 'https://open.spotify.com/track/3J...  3JJ4BoL9WVHk4Yye2EGJC7   

         name  popularity  track_number   type  
0       QKThr          63            19  track  
1        Xtal          61             1  track  
2  Avril 14th          59             9  track  
3          #3          58             3  track  
4        Flim          56             2  track

In [93]:
dict_top_tracks = df_top_tracks_clean.to_dict()

nome_top_tracks = dict_top_tracks['artists'][0][0]['name']
df_top_tracks_clean['external_urls'] = dict_top_tracks['artists'][0][0]['external_urls']['spotify']
df_top_tracks_clean['uri'] = uri = dict_top_tracks['artists'][0][0]['uri']
df_top_tracks_clean['nome'] = nome_top_tracks
df_top_tracks_processed = df_top_tracks_clean.drop(['artists'], axis=1)


In [94]:
df_top_tracks_processed

available_markets  duration_ms  explicit  \
0  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...        87466     False   
1  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       293751     False   
2  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       125520     False   
3  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       464293     False   
4  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       177093     False   
5  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       321880     False   
6  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       228053     False   
7  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       614592     False   
8  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...        77346     False   
9  [AR, AU, AT, BE, BO, BR, BG, CA, CL, CO, CR, C...       323186     False   

                                       external_urls                      id  \
0  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  4LIM4qmpHABufePRrLWbiM   
1  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  7o2AeQZzfCERsRmOM86EcB   
2  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  1uaGSDFsLdReQgg8p7Obwh   
3  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  7glKwbR1DyuIuE6XvZvJbQ   
4  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  3JJ4BoL9WVHk4Yye2EGJC7   
5  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  5ljMlD10En5rRGZU0cs2Np   
6  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  643gyipSU7dkmrFhJ8UAIm   
7  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  6gBbPTmT3qveiYpqAYvYJH   
8  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  2FQviMkaoixlgWuGL4u1EO   
9  https://open.spotify.com/artist/6kBDZFXuLrZgHn...  7KRQoq9GeWeCm0ZAXg5XMb   

              name  popularity  track_number   type  \
0            QKThr          63            19  track   
1             Xtal          61             1  track   
2       Avril 14th          59             9  track   
3               #3          58             3  track   
4             Flim          56             2  track   
5  aisatsana [102]          55            12  track   
6       Pulsewidth          54             3  track   
7              #19          45            19  track   
8                I          53             5  track   
9       Ageispolis          52             4  track   

                                     uri        nome  
0  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
1  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
2  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
3  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
4  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
5  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
6  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
7  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
8  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin  
9  spotify:artist:6kBDZFXuLrZgHnvmPu9NsG  Aphex Twin

In [ ]:
#

In [126]:
top_tracks_csv = "top_tracks_processed.csv"
albums_csv = "albums_Aphex_processed.csv"
all_tracks_csv = "all_tracks_processed.csv"

df_top_tracks_processed.to_csv(top_tracks_csv, index=False)
df_albums_Aphex_processed.to_csv(albums_csv, index=False)
df_all_tracks_processed.to_csv(all_tracks_csv, index=False)

csvs = [all_tracks_csv, albums_csv, top_tracks_csv]
destino = r'C:\ETL\data\processed'
for csv in csvs:
    source = os.path.join(os.getcwd(), os.path.basename(csv))
    destination = os.path.join(destino, os.path.basename(csv))
    if os.path.exists(destination):  
        print(f"Arquivo {csv} já existe em {destino}, não será movido.")
        os.remove(source) #quando rodava mais de uma vez, o arquivo que não era movido ficava dentro da pasta de origem (essa aqui), então resolvi excluir, acho que se apenas movesse denovo o arquivo csv gerado iria sobrescrever o antigo (que ja estava na pasta de destino)
    else:
        shutil.move(source, destination)
        print(f"Arquivo {csv} movido para {destino}")



Arquivo all_tracks_processed.csv já existe em C:\ETL\data\processed, não será movido.
Arquivo albums_Aphex_processed.csv já existe em C:\ETL\data\processed, não será movido.
Arquivo top_tracks_processed.csv já existe em C:\ETL\data\processed, não será movido.


top_tracks_processed2.csv
